File: ios/Plugin/IntercomPlugin.swift
import Foundation
import Capacitor
import Intercom

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(IntercomPlugin)
public class IntercomPlugin: CAPPlugin {
    private var observers: [NSObjectProtocol] = []

    override public func load() {
        let apiKey = getConfig().getString("iosApiKey") ?? "ADD_IN_CAPACITOR_CONFIG_JSON"
        let appId = getConfig().getString("iosAppId") ?? "ADD_IN_CAPACITOR_CONFIG_JSON"
        Intercom.setApiKey(apiKey, forAppId: appId)

        #if DEBUG
        Intercom.enableLogging()
        #endif

        NotificationCenter.default.addObserver(self, selector: #selector(self.didRegisterWithToken(notification:)), name: Notification.Name.capacitorDidRegisterForRemoteNotifications, object: nil)

        observers.append(
            NotificationCenter.default.addObserver(forName: .IntercomWindowDidShow, object: nil, queue: OperationQueue.main) { [weak self] (_) in
            self?.notifyListeners("windowDidShow", data: nil)
        })

        observers.append(NotificationCenter.default.addObserver(forName: .IntercomWindowDidHide, object: nil, queue: OperationQueue.main) { [weak self] (_) in
            self?.notifyListeners("windowDidHide", data: nil)
        })

    }

    deinit {
        NotificationCenter.default.removeObserver(self)
        for observer in observers {
            NotificationCenter.default.removeObserver(observer)
        }
    }

    @objc func didRegisterWithToken(notification: NSNotification) {
        guard let deviceToken = notification.object as? Data else {
            return
        }
        Intercom.setDeviceToken(deviceToken)
    }
    
    @objc func loadWithKeys(_ call: CAPPluginCall) {
        let appId = call.getString("appId") ?? "NO_APP_ID_PASSED"
        let apiKey = call.getString("apiKeyIOS") ?? "NO_API_KEY_PASSED"

        Intercom.setApiKey(apiKey, forAppId: appId)

        NotificationCenter.default.addObserver(self, selector: #selector(self.didRegisterWithToken(notification:)), name: Notification.Name.capacitorDidRegisterForRemoteNotifications, object: nil)
    }

    @objc func registerIdentifiedUser(_ call: CAPPluginCall) {
        let userId = call.getString("userId")
        let email = call.getString("email")
        let attributes = ICMUserAttributes()

        if (email) != nil {
            attributes.email = email
            Intercom.loginUser(with: attributes) { result in
                switch result {
                case .success: call.resolve()
                case .failure(let error): call.reject("Error logging in: \(error.localizedDescription)")
                }
            }
        }
        else if (userId) != nil {
            attributes.userId = userId
            Intercom.loginUser(with: attributes) { result in
                switch result {
                case .success: call.resolve()
                case .failure(let error): call.reject("Error logging in: \(error.localizedDescription)")
                }
            }
        }
    }

    @objc func registerUnidentifiedUser(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.loginUnidentifiedUser()
            call.resolve()
        }
    }
    
    @objc func isUserLoggedIn(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            call.resolve(["isLoggedIn": Intercom.isUserLoggedIn() ])
        }
    }

    @objc func updateUser(_ call: CAPPluginCall) {
        let userAttributes = ICMUserAttributes()
        
        if let userId = call.getString("userId") {
            userAttributes.userId = userId
        }
        if let email = call.getString("email") {
            userAttributes.email = email
        }
        if let name = call.getString("name") {
            userAttributes.name = name
        }
        if let phone = call.getString("phone") {
            userAttributes.phone = phone
        }
        if let languageOverride = call.getString("languageOverride") {
            userAttributes.languageOverride = languageOverride
        }
        if let customAttributes = call.getObject("customAttributes") {
            userAttributes.customAttributes = customAttributes
        }
        if let company = call.getObject("company") {
            let companyAttributes = ICMCompany()
            
            if let companyId = company["companyId"] as? String {
                companyAttributes.companyId = companyId
            }
            if let name = company["name"] as? String {
                companyAttributes.name = name
            }
            if let createdAt = company["createdAt"] as? TimeInterval {
                companyAttributes.createdAt = Date(timeIntervalSince1970: createdAt)
            }
            if let companyCustomAttributes = company["customAttributes"] as? JSObject {
                companyAttributes.customAttributes = companyCustomAttributes
            }
            
            userAttributes.companies = [companyAttributes]
        }
        
        DispatchQueue.main.async {
            Intercom.updateUser(with: userAttributes)
            call.resolve()
        }
    }

    @objc func logout(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.logout()
            call.resolve()
        }
    }

    @objc func logEvent(_ call: CAPPluginCall) {
        let eventName = call.getString("name")
        let metaData = call.getObject("data")

        if eventName != nil && metaData != nil {
            Intercom.logEvent(withName: eventName!, metaData: metaData!)
            
        } else if (eventName != nil) {
            Intercom.logEvent(withName: eventName!)
        }

        call.resolve()
    }

    @objc func displayMessenger(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.present();
            call.resolve()
        }
    }

    @objc func displayMessageComposer(_ call: CAPPluginCall) {
        guard let initialMessage = call.getString("message") else {
            call.reject("Enter an initial message")
            return
        }
        
        DispatchQueue.main.async {
            Intercom.presentMessageComposer(initialMessage);
            call.resolve()
        }
    }

    @objc func displayHelpCenter(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.present(.helpCenter)
            call.resolve()
        }
    }

    @objc func hideMessenger(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.hide()
            call.resolve()
        }
    }

    @objc func displayLauncher(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.setLauncherVisible(true)
            call.resolve()
        }
    }

    @objc func hideLauncher(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.setLauncherVisible(false)
            call.resolve()
        }
    }

    @objc func displayInAppMessages(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.setInAppMessagesVisible(true)
            call.resolve()
        }
    }

    @objc func hideInAppMessages(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.setInAppMessagesVisible(false)
            call.resolve()
        }
    }

    @objc func displayCarousel(_ call: CAPPluginCall) {
        guard let carouselId = call.getString("carouselId") else {
            call.reject("carouselId not provided to displayCarousel.")
            return
        }
        let carouselToPresent = Intercom.Content.carousel(id: carouselId)

        DispatchQueue.main.async {
            Intercom.presentContent(carouselToPresent)
            call.resolve()
        }
    }

    @objc func setUserHash(_ call: CAPPluginCall) {
        guard let hmac = call.getString("hmac") else {
            call.reject("No hmac found. Read intercom docs and generate it.")
            return
        }

        DispatchQueue.main.async {
            Intercom.setUserHash(hmac)
            call.resolve()
        }
    }

    @objc func setBottomPadding(_ call: CAPPluginCall) {
        guard let value = call.getString("value") else {
            call.reject("Enter a value for padding bottom")
            return;
        }
        
        if let number = NumberFormatter().number(from: value) {
            Intercom.setBottomPadding(CGFloat(truncating: number))
            call.resolve()
        } else {
            call.reject("Enter a numeric value for padding bottom")
        }
    }

    @objc func displayArticle(_ call: CAPPluginCall) {
        if let articleId = call.getString("articleId") {
            let articleToPresent = Intercom.Content.article(id: articleId)
            Intercom.presentContent(articleToPresent)
            call.resolve()
        } else {
            call.reject("articleId not provided to presentArticle.")
        }
    }
}

File: android/src/main/java/com/getcapacitor/community/intercom/intercom/IntercomPlugin.java
package com.getcapacitor.community.intercom.intercom;

import androidx.annotation.NonNull;
import com.getcapacitor.CapConfig;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import io.intercom.android.sdk.Company;
import io.intercom.android.sdk.Intercom;
import io.intercom.android.sdk.IntercomContent;
import io.intercom.android.sdk.IntercomError;
import io.intercom.android.sdk.IntercomPushManager;
import io.intercom.android.sdk.IntercomSpace;
import io.intercom.android.sdk.IntercomStatusCallback;
import io.intercom.android.sdk.UserAttributes;
import io.intercom.android.sdk.identity.Registration;
import io.intercom.android.sdk.push.IntercomPushClient;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import org.json.JSONException;

@CapacitorPlugin(name = "Intercom", permissions = @Permission(strings = {}, alias = "receive"))
public class IntercomPlugin extends Plugin {

    private static final String EVENT_WINDOW_DID_SHOW = "windowDidShow";
    private static final String EVENT_WINDOW_DID_HIDE = "windowDidHide";

    private final IntercomPushClient intercomPushClient = new IntercomPushClient();

    @Override
    public void load() {
        // Set up Intercom
        setUpIntercom();

        // load parent
        super.load();
    }

    @PluginMethod
    public void loadWithKeys(PluginCall call) {
        String appId = call.getString("appId", "NO_APP_ID_PASSED");
        String apiKey = call.getString("apiKeyAndroid", "NO_API_KEY_PASSED");

        Intercom.initialize(this.getActivity().getApplication(), apiKey, appId);

        // load parent
        super.load();
    }

    @Override
    public void handleOnStart() {
        super.handleOnStart();
        bridge
            .getActivity()
            .runOnUiThread(
                new Runnable() {
                    @Override
                    public void run() {
                        //We also initialize intercom here just in case it has died. If Intercom is already set up, this won't do anything.
                        setUpIntercom();
                        Intercom.client().handlePushMessage();
                    }
                }
            );
    }

    @Override
    public void handleOnPause() {
        super.handleOnPause();
        notifyListeners(EVENT_WINDOW_DID_SHOW, null);
    }

    @Override
    public void handleOnResume() {
        super.handleOnResume();
        notifyListeners(EVENT_WINDOW_DID_HIDE, null);
    }

    @PluginMethod
    public void registerIdentifiedUser(PluginCall call) {
        String email = call.getString("email");
        String userId = call.getData().getString("userId");

        Registration registration = new Registration();

        if (email != null && email.length() > 0) {
            registration = registration.withEmail(email);
        }
        if (userId != null && userId.length() > 0) {
            registration = registration.withUserId(userId);
        }

        Intercom.client().loginIdentifiedUser(registration, new IntercomStatusCallback() {
            @Override
            public void onSuccess() {
                call.resolve();
            }

            @Override
            public void onFailure(@NonNull IntercomError intercomError) {
                call.reject(intercomError.getErrorMessage());
            }
        });
    }

    @PluginMethod
    public void registerUnidentifiedUser(PluginCall call) {
        Intercom.client().loginUnidentifiedUser(new IntercomStatusCallback() {
            @Override
            public void onSuccess() {
                call.resolve();
            }

            @Override
            public void onFailure(@NonNull IntercomError intercomError) {
                call.reject(intercomError.getErrorMessage());
            }
        });
    }

    @PluginMethod
    public void updateUser(PluginCall call) throws JSONException {
        UserAttributes.Builder builder = new UserAttributes.Builder();
        String userId = call.getString("userId");
        if (userId != null && userId.length() > 0) {
            builder.withUserId(userId);
        }
        String email = call.getString("email");
        if (email != null && email.length() > 0) {
            builder.withEmail(email);
        }
        String name = call.getString("name");
        if (name != null && name.length() > 0) {
            builder.withName(name);
        }
        String phone = call.getString("phone");
        if (phone != null && phone.length() > 0) {
            builder.withPhone(phone);
        }
        String languageOverride = call.getString("languageOverride");
        if (languageOverride != null && languageOverride.length() > 0) {
            builder.withLanguageOverride(languageOverride);
        }
        Map<String, Object> customAttributes = mapFromJSON(call.getObject("customAttributes"));
        builder.withCustomAttributes(customAttributes);

        JSObject company = call.getObject("company");
        if (company != null) {
            Company.Builder companyBuilder = new Company.Builder();
            String companyId = company.getString("companyId");
            if (companyId != null) {
                companyBuilder.withCompanyId(companyId);
            }
            String companyName = company.getString("name");
            if (companyName != null) {
                companyBuilder.withName(companyName);
            }

            Map<String, Object> companyCustomAttributes = mapFromJSON(company.getJSObject("customAttributes"));
            if (companyCustomAttributes != null) {
                companyBuilder.withCustomAttributes(companyCustomAttributes);
            }

            try {
                Long createdAt = company.getLong("createdAt");
                companyBuilder.withCreatedAt(createdAt);
            } catch (Exception e) {
                Logger.error("Intercom", "ERROR: Could not handle company.createdAt", e);
            }

            builder.withCompany(companyBuilder.build());
        }

        Intercom.client().updateUser(builder.build(),  new IntercomStatusCallback() {
            @Override
            public void onSuccess() {
                call.resolve();
            }

            @Override
            public void onFailure(@NonNull IntercomError intercomError) {
                call.reject(intercomError.getErrorMessage());
            }
        });
    }

    @PluginMethod
    public void isUserLoggedIn(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("isLoggedIn", Intercom.client().isUserLoggedIn());
        call.resolve(ret);
    }

    @PluginMethod
    public void logout(PluginCall call) {
        Intercom.client().logout();
        call.resolve();
    }

    @PluginMethod
    public void logEvent(PluginCall call) {
        String eventName = call.getString("name");
        Map<String, Object> metaData = mapFromJSON(call.getObject("data"));
        if (eventName != null) {
            if (metaData == null) {
                Intercom.client().logEvent(eventName);
            } else {
                Intercom.client().logEvent(eventName, metaData);
            }
            call.resolve();
        } else {
            call.reject("logEvent missing 'name' parameter");
        }
    }

    @PluginMethod
    public void displayMessenger(PluginCall call) {
        Intercom.client().present(IntercomSpace.Home);
        call.resolve();
    }

    @PluginMethod
    public void displayMessageComposer(PluginCall call) {
        String message = call.getString("message");
        Intercom.client().displayMessageComposer(message);
        call.resolve();
    }

    @PluginMethod
    public void displayHelpCenter(PluginCall call) {
        Intercom.client().present(IntercomSpace.HelpCenter);
        call.resolve();
    }

    @PluginMethod
    public void hideMessenger(PluginCall call) {
        Intercom.client().hideIntercom();
        call.resolve();
    }

    @PluginMethod
    public void displayLauncher(PluginCall call) {
        Intercom.client().setLauncherVisibility(Intercom.VISIBLE);
        call.resolve();
    }

    @PluginMethod
    public void hideLauncher(PluginCall call) {
        Intercom.client().setLauncherVisibility(Intercom.GONE);
        call.resolve();
    }

    @PluginMethod
    public void displayInAppMessages(PluginCall call) {
        Intercom.client().setInAppMessageVisibility(Intercom.VISIBLE);
        call.resolve();
    }

    @PluginMethod
    public void hideInAppMessages(PluginCall call) {
        Intercom.client().setLauncherVisibility(Intercom.GONE);
        call.resolve();
    }

    @PluginMethod
    public void displayCarousel(PluginCall call) {
        String carouselId = call.getString("carouselId");
        if (carouselId == null) {
            call.reject("Missing displayCarousel 'carouselId'");
        } else {
            Intercom.client().presentContent(new IntercomContent.Carousel(carouselId));
            call.resolve();
        }
    }

    @PluginMethod
    public void setUserHash(PluginCall call) {
        String hmac = call.getString("hmac");
        if (hmac == null) {
            call.reject("Missing setUserHash 'hmac'");
        } else {
            Intercom.client().setUserHash(hmac);
            call.resolve();
        }
    }

    @PluginMethod
    public void setBottomPadding(PluginCall call) {
        String stringValue = call.getString("value");
        if (stringValue == null) {
            call.reject("Missing bottom padding 'value'");
        } else {
            int value = Integer.parseInt(stringValue);
            Intercom.client().setBottomPadding(value);
            call.resolve();
        }
    }

    @PluginMethod
    public void sendPushTokenToIntercom(PluginCall call) {
        String token = call.getString("value");

        if (token == null) {
            call.reject("sendPushTokenToIntercom missing 'value'");
        } else {
            try {
                intercomPushClient.sendTokenToIntercom(this.getActivity().getApplication(), token);
                call.resolve();
            } catch (Exception e) {
                call.reject("Failed to send push token to Intercom", e);
            }
        }
    }

    @PluginMethod
    public void receivePush(PluginCall call) {
        try {
            JSObject notificationData = call.getData();
            Map message = mapFromJSON(notificationData);
            if (intercomPushClient.isIntercomPush(message)) {
                intercomPushClient.handlePush(this.getActivity().getApplication(), message);
                call.resolve();
            } else {
                call.reject("Notification data was not a valid Intercom push message");
            }
        } catch (Exception e) {
            call.reject("Failed to handle received Intercom push", e);
        }
    }

    @PluginMethod
    public void displayArticle(PluginCall call) {
        String articleId = call.getString("articleId");
        if (articleId == null) {
            call.reject("displayArticle missing 'articleId");
        } else {
            Intercom.client().presentContent(new IntercomContent.Article((articleId)));
            call.resolve();
        }
    }

    private void setUpIntercom() {
        try {
            // get config
            CapConfig config = this.bridge.getConfig();
            String apiKey = config.getPluginConfiguration("Intercom").getString("androidApiKey");
            String appId = config.getPluginConfiguration("Intercom").getString("androidAppId");

            // init intercom sdk
            Intercom.initialize(this.getActivity().getApplication(), apiKey, appId);
        } catch (Exception e) {
            Logger.error("Intercom", "ERROR: Something went wrong when initializing Intercom. Check your configurations", e);
        }
    }

    private static Map<String, Object> mapFromJSON(JSObject jsonObject) {
        if (jsonObject == null) {
            return null;
        }
        Map<String, Object> map = new HashMap<>();
        Iterator<String> keysIter = jsonObject.keys();
        while (keysIter.hasNext()) {
            String key = keysIter.next();
            Object value = getObject(jsonObject.opt(key));
            if (value != null) {
                map.put(key, value);
            }
        }
        return map;
    }

    private static Object getObject(Object value) {
        if (value instanceof JSObject) {
            value = mapFromJSON((JSObject) value);
        } else if (value instanceof JSArray) {
            value = listFromJSON((JSArray) value);
        }
        return value;
    }

    private static List<Object> listFromJSON(JSArray jsonArray) {
        List<Object> list = new ArrayList<>();
        for (int i = 0, count = jsonArray.length(); i < count; i++) {
            Object value = getObject(jsonArray.opt(i));
            if (value != null) {
                list.add(value);
            }
        }
        return list;
    }
}
